### ISSUE TITLE
Number of squares and multiplications

### ISSUE THREAD

**Issue by kevaundray** [mmcloughlin/addchain#54]
Perhaps add the number of squares and multiplications or doublings and additions alongside the chain length?

**Comment by mmcloughlin** [comment]
To clarify, are you talking about the results lists (readme and [full results page](https://github.com/mmcloughlin/addchain/blob/master/doc/results.md))?

**Comment by kevaundray** [comment]
Yep, the readme list.

For the program, I haven’t looked around, so I am not sure how hard it would be to in-cooperate the fact that squaring is usually cheaper than multiplying or whether it would be a useful heuristic?

**Comment by mmcloughlin** [comment]
Yes, I agree with you on showing the add/double counts (equivalent to multiplication/square). It's common for other work on the subject to present it that way too, for example https://eprint.iacr.org/2018/1038.

When you say incorporate, do you mean use a weighting when selecting the best chain from a bunch of results?

**Comment by kevaundray** [comment]
Yep, I mean to use a weighting. For example, setting the weighting of a squaring operation to be 0.85 the weighting of a multiplication operation.

One possible drawback, if it has utility, is that the  actual weighting is implementation specific, so a blanket ratio may not be suitable.

**Comment by mmcloughlin** [comment]
> Yep, I mean to use a weighting. For example, setting the weighting of a squaring operation to be 0.85 the weighting of a multiplication operation.

Right. I considered this a while back and forgot about it, so it didn't make it into the initial release, but I agree this would be a good feature. The fact that squares are cheaper is accounted for to some extent here:

https://github.com/mmcloughlin/addchain/blob/fe56d6780fb9b9417a1d7c2e7535d66b6effdfb5/internal/results/results_test.go#L49-L65

That is, among the shortest chains it picks the one with the fewest adds (equivalently the most doubles). After that it sorts by algorithm name... that was just a way to ensure the result was deterministic!

> One possible drawback, if it has utility, is that the  actual weighting is implementation specific, so a blanket ratio may not be suitable.

Yes, agreed the exact weightings would need to be configurable.

I think there's an interesting question about what other heuristics you would use to choose the best chain, even among those that have the exact same number of adds and doubles. @briansmith in some private communication suggested both maximum number of live variables and pipeline stalls (dependencies between variables) as factors you might want to consider.

As long as you have some coarse score to narrow down to the top candidates, then you could just code generate the actual exponentiation function and benchmark it.

**Comment by kevaundray** [comment]
I think the dependency between variables is a good one, it would be interesting to see to what extent this heuristic speeds up SIMD implementations, if at all.
